[Java] Added section for converting Avro To Arrow (#235)
apache/arrow-cookbook: java/source/avro.rst

.. Licensed to the Apache Software Foundation (ASF) under one
.. or more contributor license agreements.  See the NOTICE file
.. distributed with this work for additional information
.. regarding copyright ownership.  The ASF licenses this file
.. to you under the Apache License, Version 2.0 (the
.. "License"); you may not use this file except in compliance
.. with the License.  You may obtain a copy of the License at

..   http://www.apache.org/licenses/LICENSE-2.0

.. Unless required by applicable law or agreed to in writing,
.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

.. _arrow-avro:

===============================
Avro to Arrow
===============================

Avro encoded data can be converted into Arrow format.

.. contents::

Avro to Arrow
=============

The example assumes that the avro data doesn't contain schema in the file
to be converted into Arrow

.. testcode::

>     import org.apache.arrow.AvroToArrow;
>     import org.apache.arrow.AvroToArrowConfig;
>     import org.apache.arrow.AvroToArrowConfigBuilder;
>     import org.apache.arrow.AvroToArrowVectorIterator;
>     import org.apache.arrow.memory.BufferAllocator;
>     import org.apache.arrow.memory.RootAllocator;
>     import org.apache.arrow.vector.VectorSchemaRoot;
>     import org.apache.avro.Schema;
>     import org.apache.avro.io.BinaryDecoder;
>     import org.apache.avro.io.DecoderFactory;
> 
>     import java.io.File;
>     import java.io.FileInputStream;
> 
>     BufferAllocator allocator = new RootAllocator(Long.MAX_VALUE);
>     AvroToArrowConfig config = new AvroToArrowConfigBuilder(allocator).build();
> 
>     BinaryDecoder decoder = new DecoderFactory().binaryDecoder(new FileInputStream("./thirdpartydeps/avro/users.avro"), null);
> 
>     Schema schema = new Schema.Parser().parse(new File("./thirdpartydeps/avro/user.avsc"));
>     AvroToArrowVectorIterator avroToArrowVectorIterator = AvroToArrow.avroToArrowIterator(schema, decoder, config);
> 
>     while(avroToArrowVectorIterator.hasNext()) {
>         VectorSchemaRoot root = avroToArrowVectorIterator.next();
>         System.out.print(root.contentToTSVString());
>     }

Reviewer: ```suggestion
    import org.apache.arrow.AvroToArrow;
    import org.apache.arrow.AvroToArrowConfig;
    import org.apache.arrow.AvroToArrowConfigBuilder;
    import org.apache.arrow.AvroToArrowVectorIterator;
    import org.apache.arrow.memory.BufferAllocator;
    import org.apache.arrow.memory.RootAllocator;
    import org.apache.arrow.vector.VectorSchemaRoot;
    import org.apache.avro.Schema;
    import org.apache.avro.io.BinaryDecoder;
    import org.apache.avro.io.DecoderFactory;

    import java.io.File;
    import java.io.FileInputStream;
    import java.io.FileNotFoundException;
    import java.io.IOException;

    try {
        BinaryDecoder decoder = new DecoderFactory().binaryDecoder(new FileInputStream("./thirdpartydeps/avro/users.avro"), null);
        Schema schema = new Schema.Parser().parse(new File("./thirdpartydeps/avro/user.avsc"));
        try (BufferAllocator allocator = new RootAllocator()) {
            AvroToArrowConfig config = new AvroToArrowConfigBuilder(allocator).build();
            try (AvroToArrowVectorIterator avroToArrowVectorIterator = AvroToArrow.avroToArrowIterator(schema, decoder, config)) {
                while(avroToArrowVectorIterator.hasNext()) {
                    try (VectorSchemaRoot root = avroToArrowVectorIterator.next()) {
                        System.out.print(root.contentToTSVString());
                    }
                }
            }
        }
    } catch (FileNotFoundException e) {
        e.printStackTrace();
    } catch (IOException e) {
        e.printStackTrace();
    }
```

Author: @davisusanibar This issue seems to be common theme from my side. How do I ensure I don't make the same mistake again?

.. Licensed to the Apache Software Foundation (ASF) under one
.. or more contributor license agreements.  See the NOTICE file
.. distributed with this work for additional information
.. regarding copyright ownership.  The ASF licenses this file
.. to you under the Apache License, Version 2.0 (the
.. "License"); you may not use this file except in compliance
.. with the License.  You may obtain a copy of the License at

..   http://www.apache.org/licenses/LICENSE-2.0

.. Unless required by applicable law or agreed to in writing,
.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

.. _arrow-avro:

===============================
Avro to Arrow
===============================

Avro encoded data can be converted into Arrow format.

.. contents::

Avro to Arrow
=============

The example assumes that the avro data doesn't contain schema in the file

> to be converted into Arrow

Reviewer 2: I honestly do not know what this means? The code sample is reading a schema it appears

Reviewer 2: Ah, separate files.

```suggestion
The example assumes that the Avro schema is stored separately from the Avro data itself.
```

Author: Arrow data file can store schema along with data. This is in reference to that. This example shall fail in that case, hence mentioned explicitly.

    import org.apache.arrow.memory.RootAllocator;
    import org.apache.arrow.vector.VectorSchemaRoot;
    import org.apache.avro.Schema;
    import org.apache.avro.io.BinaryDecoder;
    import org.apache.avro.io.DecoderFactory;

    import java.io.File;
    import java.io.FileInputStream;
    import java.io.FileNotFoundException;
    import java.io.IOException;

    try {
        BinaryDecoder decoder = new DecoderFactory().binaryDecoder(new FileInputStream("./thirdpartydeps/avro/users.avro"), null);
        Schema schema = new Schema.Parser().parse(new File("./thirdpartydeps/avro/user.avsc"));
        try (BufferAllocator allocator = new RootAllocator()) {
            AvroToArrowConfig config = new AvroToArrowConfigBuilder(allocator).build();
            try (AvroToArrowVectorIterator avroToArrowVectorIterator = AvroToArrow.avroToArrowIterator(schema, decoder, config)) {
                while(avroToArrowVectorIterator.hasNext()) {
                    try (VectorSchemaRoot root = avroToArrowVectorIterator.next()) {
                        System.out.print(root.contentToTSVString());
                    }
                }
            }
        }

>     } catch (FileNotFoundException e) {

Reviewer 2: Consolidate the catches?

Author: Done

Reviewer 2: It doesn't appear to have changed, additionally the code and output are still misaligned by one space

Author: I am just following format similar to other files, hence didn't remove extra space char. Should I align? Should we do it for other files too?

Reviewer 2: Oops. It should generally be aligned.

Author: Let me go ahead and fix the alignment, if it looks good we shall take care in other files when we touch them. Give me 5 mins

Author: pushed the change, one issue i noted it the html rendered output has extra space between results, not sure where it is coming from. Here is how it looks in html

name    favorite_number    favorite_color
Alyssa    256    null

Ben    7    red
